File: sdks/python/src/e2a/v1/_retry.py
"""Retry + idempotency for the e2a Python SDK (Slice 8c).

The TypeScript SDK wraps the generated HTTP library to retry the *same* request
(idempotency key minted once, bytes reused). The generated Python (httpx) base
exposes no such seam — its ``ApiClient`` builds its own client — so we retry at
the hand-written resource layer instead:

* Mint an ``Idempotency-Key`` ONCE per write and pass it (via ``_headers``) on
  every attempt, so the server dedupes a retried send/reply/forward/approve.
* Retry only the SAFE subset on a retryable failure: reads, the server-deduped
  keyed writes, and HTTP-idempotent writes (PUT/PATCH/DELETE). A bare,
  non-idempotent POST (create/reject/redeliver/test) is NOT retried on an
  ambiguous failure — re-sending could double-create.
"""

from __future__ import annotations

import asyncio
import random as _random
import time as _time
import uuid
from typing import Any, Awaitable, Callable, Optional

import httpx

from .errors import (
    E2AError,
    connection_error,
    from_api_exception,
    is_retryable_status,
)
from .generated.exceptions import ApiException

# make_coro receives the per-attempt extra headers (the minted Idempotency-Key,
# or None for reads) and returns the awaitable generated call.
MakeCoro = Callable[[Optional["dict[str, str]"]], Awaitable[Any]]

IDEMPOTENCY_HEADER = "Idempotency-Key"


class RetryConfig:
    """Tunables (and test seams) for :func:`request_with_retry`."""

    def __init__(
        self,
        *,
        max_retries: int = 2,
        base_delay_ms: float = 200.0,
        max_delay_ms: float = 8_000.0,
        max_retry_after_ms: float = 60_000.0,
        max_elapsed_ms: Optional[float] = None,
        sleep: Callable[[float], Awaitable[None]] = asyncio.sleep,
        rand: Callable[[], float] = _random.random,
        now: Callable[[], float] = _time.monotonic,
        gen_idempotency_key: Callable[[], str] = lambda: uuid.uuid4().hex,
    ) -> None:
        self.max_retries = max_retries
        self.base_delay_ms = base_delay_ms
        self.max_delay_ms = max_delay_ms
        self.max_retry_after_ms = max_retry_after_ms
        self.max_elapsed_ms = max_elapsed_ms
        self.sleep = sleep
        self.rand = rand
        self.now = now
        self.gen_idempotency_key = gen_idempotency_key


def _backoff_ms(cfg: RetryConfig, attempt: int, exc: Optional[ApiException]) -> float:
    # Honor Retry-After (seconds) up to max_retry_after_ms; else exp + jitter.
    if exc is not None:
        retry_after = _retry_after_ms(exc)
        if retry_after is not None:
            return min(retry_after, cfg.max_retry_after_ms)
    exp = min(cfg.base_delay_ms * (2 ** attempt), cfg.max_delay_ms)
    return exp * (0.5 + 0.5 * cfg.rand())  # full jitter over [0.5, 1.0]


def _retry_after_ms(exc: ApiException) -> Optional[float]:
    headers = getattr(exc, "headers", None)
    if not headers:
        return None
    try:
        items = headers.items()
    except AttributeError:
        return None
    for k, v in items:
        if k.lower() == "retry-after":
            try:
                secs = float(v)
                if secs >= 0:
                    return secs * 1000.0
            except (TypeError, ValueError):
                return None
    return None


async def request_with_retry(
    make_coro: MakeCoro,
    *,
    cfg: RetryConfig,
    retryable: bool,
    idempotency: bool,
    idempotency_key: Optional[str] = None,
) -> Any:
    """Execute a generated call with retry + idempotency.

    :param retryable: whether a retryable failure (429/5xx/connection) should be
        retried. False for non-idempotent POSTs.
    :param idempotency: whether to mint/forward an Idempotency-Key header.
    :param idempotency_key: caller-supplied stable key; minted if omitted.
    """
    headers: Optional["dict[str, str]"] = None
    if idempotency:
        key = idempotency_key or cfg.gen_idempotency_key()
        headers = {IDEMPOTENCY_HEADER: key}

    start = cfg.now()
    attempt = 0
    while True:
        exc: Optional[BaseException] = None
        api_exc: Optional[ApiException] = None
        try:
            return await make_coro(headers)
        except ApiException as e:
            api_exc = e
            exc = e
            can_retry = retryable and is_retryable_status(int(getattr(e, "status", 0) or 0))
        except httpx.TransportError as e:  # connection-level: no HTTP response
            exc = e
            can_retry = retryable
        except E2AError:
            raise  # already typed (e.g. a nested helper) — pass through
        except httpx.HTTPError as e:
            # Non-transport httpx error (InvalidURL, etc.) — not retryable, but
            # surface it as a typed E2AError rather than a raw httpx exception.
            raise connection_error(str(e), cause=e)

        if not can_retry or attempt >= cfg.max_retries:
            raise _as_e2a_error(exc, api_exc)

        delay_ms = _backoff_ms(cfg, attempt, api_exc)
        if cfg.max_elapsed_ms is not None and (cfg.now() - start) * 1000.0 + delay_ms > cfg.max_elapsed_ms:
            raise _as_e2a_error(exc, api_exc)
        await cfg.sleep(delay_ms / 1000.0)
        attempt += 1


def _as_e2a_error(exc: Optional[BaseException], api_exc: Optional[ApiException]) -> E2AError:
    if api_exc is not None:
        return from_api_exception(api_exc)
    message = str(exc) if exc is not None else "connection error"
    return connection_error(message, cause=exc)

File: sdks/typescript/src/v1/retry.ts
// Retry + idempotency for the e2a SDK (Slice 8b-1).
//
// Implemented as a wrapping HttpLibrary rather than generated middleware: middleware
// can transform a request/response but cannot re-send, and re-sending is the
// whole point. Because we re-send the SAME RequestContext, the Idempotency-Key
// header and the already-serialized body are reused verbatim across attempts —
// which is exactly what makes a POST retry safe (the server dedupes on the key
// and hashes the raw bytes). The key is minted ONCE, before the first attempt.

import { from, Observable } from "./generated/rxjsStub.js";
import { HttpMethod } from "./generated/http/http.js";
import type { HttpLibrary, RequestContext, ResponseContext } from "./generated/http/http.js";
import { isRetryableStatus } from "./errors.js";

export interface RetryOptions {
  /** Max retry attempts after the first try. Default 2. */
  maxRetries?: number;
  /** Base backoff in ms (doubles per attempt). Default 200. */
  baseDelayMs?: number;
  /** Exponential-backoff ceiling in ms. Default 8000. */
  maxDelayMs?: number;
  /** Upper bound for an honored Retry-After, in ms. Default 60000 — a hostile
   *  or buggy upstream can otherwise wedge a request for years. */
  maxRetryAfterMs?: number;
  /** Optional total deadline across all attempts (incl. backoff), in ms. When
   *  the next sleep would exceed it, stop and return/throw the last result. */
  maxElapsedMs?: number;
  /** Injectable for tests. */
  sleep?: (ms: number) => Promise<void>;
  random?: () => number;
  genIdempotencyKey?: () => string;
  now?: () => number;
}

// Unsafe methods get an auto idempotency key when the caller didn't supply one.
const UNSAFE_METHODS = new Set<HttpMethod>([
  HttpMethod.POST,
  HttpMethod.PATCH,
  HttpMethod.PUT,
  HttpMethod.DELETE,
]);

const IDEMPOTENCY_HEADER = "Idempotency-Key";

function defaultSleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function abortError(reason?: unknown): unknown {
  if (reason !== undefined) return reason;
  return new DOMException("The operation was aborted", "AbortError");
}

function defaultUuid(): string {
  // Every supported runtime (Node 18+, browsers, edge) has crypto.randomUUID;
  // the fallback is a non-crypto last resort and is effectively unreachable.
  const c = (globalThis as { crypto?: { randomUUID?: () => string } }).crypto;
  if (c?.randomUUID) return c.randomUUID();
  return `${Date.now().toString(16)}-${Math.random().toString(16).slice(2)}${Math.random().toString(16).slice(2)}`;
}

export class RetryHttpLibrary implements HttpLibrary {
  constructor(
    private readonly inner: HttpLibrary,
    private readonly opts: RetryOptions = {},
  ) {}

  send(request: RequestContext): Observable<ResponseContext> {
    return from(this.sendWithRetry(request));
  }

  private async sendWithRetry(request: RequestContext): Promise<ResponseContext> {
    this.ensureIdempotencyKey(request);
    const max = this.opts.maxRetries ?? 2;
    const now = this.opts.now ?? Date.now;
    const start = now();
    const signal = request.getSignal?.();

    let attempt = 0;
    for (;;) {
      this.throwIfAborted(signal);

      let resp: ResponseContext | undefined;
      let connErr: unknown;
      try {
        resp = await this.inner.send(request).toPromise();
      } catch (e) {
        connErr = e; // connection-level failure (no HTTP response)
      }

      const isConn = resp === undefined;
      const retryable = isConn || isRetryableStatus(resp!.httpStatusCode);
      if (!retryable || attempt >= max) {
        if (resp !== undefined) return resp;
        throw connErr;
      }

      const delay = this.backoffMs(attempt, resp);
      // Total-deadline guard: don't start a wait that would blow the deadline.
      if (this.opts.maxElapsedMs !== undefined && now() - start + delay > this.opts.maxElapsedMs) {
        if (resp !== undefined) return resp;
        throw connErr;
      }
      await this.sleep(delay, signal);
      attempt += 1;
    }
  }

  // Mint an Idempotency-Key once, before the first attempt, for an unsafe
  // method the caller didn't already key. Setting it on the shared
  // RequestContext means every retry of this request reuses the same key.
  //
  // The generated layer unconditionally calls setHeaderParam("Idempotency-Key",
  // serialize(undefined)) on send/reply/forward/approve, leaving the header
  // *present but empty* when the caller passed no key. So "present" isn't enough
  // to mean "caller-supplied" — only a non-empty value counts; otherwise we mint
  // (overwriting the empty stub).
  private ensureIdempotencyKey(request: RequestContext): void {
    if (!UNSAFE_METHODS.has(request.getHttpMethod())) return;
    const headers = request.getHeaders();
    for (const k of Object.keys(headers)) {
      if (k.toLowerCase() !== IDEMPOTENCY_HEADER.toLowerCase()) continue;
      const v = headers[k];
      if (v != null && String(v).trim() !== "") return; // genuine caller-supplied key wins
      break; // present-but-empty generated stub → fall through and mint
    }
    request.setHeaderParam(IDEMPOTENCY_HEADER, (this.opts.genIdempotencyKey ?? defaultUuid)());
  }

  private throwIfAborted(signal: AbortSignal | undefined): void {
    if (signal?.aborted) throw abortError(signal.reason);
  }

  // Backoff: honor Retry-After (seconds or HTTP-date) up to maxRetryAfterMs;
  // otherwise exponential + jitter capped at maxDelayMs.
  private backoffMs(attempt: number, resp?: ResponseContext): number {
    const ra = resp?.headers ? this.retryAfterMs(resp.headers) : undefined;
    if (ra !== undefined) return Math.min(ra, this.opts.maxRetryAfterMs ?? 60000);

    const maxDelay = this.opts.maxDelayMs ?? 8000;
    const base = this.opts.baseDelayMs ?? 200;
    const exp = Math.min(base * 2 ** attempt, maxDelay);
    const rand = (this.opts.random ?? Math.random)();
    return Math.round(exp * (0.5 + 0.5 * rand)); // full jitter over [0.5, 1.0]
  }

  private retryAfterMs(headers: Record<string, string>): number | undefined {
    for (const k of Object.keys(headers)) {
      if (k.toLowerCase() !== "retry-after") continue;
      const v = headers[k];
      const secs = Number(v);
      if (Number.isFinite(secs) && secs >= 0) return secs * 1000;
      // RFC 9110 §10.2.3 allows an HTTP-date — common behind CDNs.
      const at = Date.parse(v);
      if (Number.isFinite(at)) return Math.max(0, at - (this.opts.now ?? Date.now)());
      return undefined;
    }
    return undefined;
  }

  private async sleep(ms: number, signal: AbortSignal | undefined): Promise<void> {
    // Injected sleep (tests) — still respect a pre-aborted signal.
    if (this.opts.sleep) {
      this.throwIfAborted(signal);
      return this.opts.sleep(ms);
    }
    if (!signal) return defaultSleep(ms);
    // Race the backoff against an abort so a cancelled request stops promptly.
    return new Promise<void>((resolve, reject) => {
      if (signal.aborted) {
        reject(abortError(signal.reason));
        return;
      }
      const t = setTimeout(() => {
        signal.removeEventListener("abort", onAbort);
        resolve();
      }, ms);
      const onAbort = () => {
        clearTimeout(t);
        reject(abortError(signal.reason));
      };
      signal.addEventListener("abort", onAbort, { once: true });
    });
  }
}
